File: strings/reorganize_string.rb
# @param {String} s
# @return {String}
def reorganize_string(s)
    n = s.size

    char_count = Hash.new(0)
    s.each_char { |c| char_count[c] += 1}

    heap = char_count.to_a.sort_by { |char, count| -count }

    return "" if heap[0][1] > (n + 1) / 2

    res = Array.new(n)

    pointer = 0

    while !heap.empty?
        char, count = heap.shift
        count.times do
            res[pointer] = char
            pointer += 2
            if pointer >= n
                pointer = 1
            end
        end
    end
    
    res.join("")
end

File: strings/reorganize_string.py
from collections import Counter
from heapq import heapify, heappop
import sys

class Solution:
    def reorganizeString(self, s: str) -> str:
        n = len(s)
        str_count = Counter(s)

        pq = [(-value, key) for key, value in str_count.items()]
        heapify(pq)

        if -pq[0][0] > (n + 1) // 2:
            return ""

        res: list[str] = [None] * n

        pointer = 0

        while len(pq) > 0:
            (count, char) = heappop(pq)
            count = -count
            for _ in range(count):
                res[pointer] = char
                pointer += 2
                if pointer >= n:
                    pointer = 1
        return "".join(res)